src/librustc_resolve/lib.rs

```diff
@@ -1691,6 +1745,9 @@ pub struct Resolver<'a> {
     current_type_ascription: Vec<Span>,
 
     injected_crate: Option<Module<'a>>,
+
+    /// Used for more accurate error when using type parameters in a associated items.
+    scope: CurrentScope,
```

Comment:
Please use some other terminology than "scope", it's incredibly overloaded in resolve.
Also, could you move `enum CurrentScope` and its impls into `diagnostics.rs`?

Reply:
What would be a better name?

Reply:
`BodyOwner` / `DefOwner` maybe? (though `Def` is similarly unfortunately overloaded)